src/MerQure.Tools/Buses/Consumer.cs

```diff
@@ -0,0 +1,64 @@
+﻿using MerQure.Messages;
+using MerQure.Tools.Configurations;
+using MerQure.Tools.Messages;
+using Newtonsoft.Json;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace MerQure.Tools.Buses
+{
+    public class Consumer<T> where T : IDelivered
+    {
+        private readonly ConsumerProvider _consumerProvider;
+        public Dictionary<string, RetryInformations> RetryInformations { get; } = new Dictionary<string, RetryInformations>();
+
+        public Consumer(IMessagingService messagingService)
+        {
+            _consumerProvider = new ConsumerProvider(messagingService);
+        }
+
+        public void Consume(Channel channel, EventHandler<T> callback)
+        {
+            _consumerProvider.Get(channel).Consume((object sender, IMessagingEvent messagingEvent) =>
+            {
+                OnMessageReceived(callback, messagingEvent);
+            });
+        }
+
+        public void AcknowlegdeDeliveredMessage(Channel channel, IDelivered deliveredMessage)
+        {
+            deliveredMessage.DeliveryTag = DecodeDeliveryTag(deliveredMessage.DeliveryTag);
+            _consumerProvider.Get(channel).AcknowlegdeDeliveredMessage(deliveredMessage);
+            RetryInformations.Remove(deliveredMessage.DeliveryTag);
+        }
+
+        public void RejectDeliveredMessage(Channel channel, IDelivered deliveredMessage)
+        {
+            deliveredMessage.DeliveryTag = DecodeDeliveryTag(deliveredMessage.DeliveryTag);
+            _consumerProvider.Get(channel).RejectDeliveredMessage(deliveredMessage);
+            RetryInformations.Remove(deliveredMessage.DeliveryTag);
+        }
+
+        public void OnMessageReceived(EventHandler<T> callback, IMessagingEvent messagingEvent)
+        {
+            RetryMessage<T> retryMessage = JsonConvert.DeserializeObject<RetryMessage<T>>(messagingEvent.Message.GetBody());
+            retryMessage.OriginalMessage.DeliveryTag = EncodeDeliveryTag(messagingEvent.DeliveryTag);
+            RetryInformations.Add(retryMessage.OriginalMessage.DeliveryTag, retryMessage.RetryInformations);
+
+            callback(this, retryMessage.OriginalMessage);
+        }
+
+        private static string EncodeDeliveryTag(string deliveryTag) //TODO CLEAN !! this is just a fast fix ...
```

Comment:
![INFO](https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO') Complete the task associated to this 'TODO' comment. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=csharpsquid%3AS1135)